### PR TITLE
Ensure Subscription Initialization is only run once

### LINKED ISSUE #5615: With UPE enabled, customers are being charged multiple times for renewal order
### Describe the bug

On the latest version of WC Payments (5.5 - released last week), subscription customers are being charged multiple times for renewal orders. 

The cause of this was introduced in this PR:

- https://github.com/Automattic/woocommerce-payments/pull/3968

As part of that PR, multiple instances of the payment method classes (eg `CC_Payment_Gateway`) are instantiated in each request. Those payment method classes extend `WC_Payment_Gateway_WCPay` and call that base class constructor. That class hierarchy leads to multiple calls of [`$this->maybe_init_subscriptions()`](https://github.com/automattic/woocommerce-payments/blob/trunk/includes/class-wc-payment-gateway-wcpay.php#L386) which hooks onto the `'woocommerce_scheduled_subscription_payment_' . $this->id` hook each time ([code ref](https://github.com/automattic/woocommerce-payments/blob/5.5.0/includes/compat/subscriptions/trait-wc-payment-gateway-wcpay-subscriptions.php#L138)). 

For as many `'woocommerce_scheduled_subscription_payment_' . $this->id` functions that are hooked on, there are that many charge requests for renewal orders. For example, with legacy UPE this was 2 (CC + UPE). But with UPE SPlit enabled there were like 8 renewal order charges (CC + all the UPE ones). 

### To Reproduce

1. Enable WC Payments 5.5.0
2.Enable UPE.
3. Enable the WC Subscriptions extension.
_3. Optional. Create a subscription product if you don't have one._ 
4. Purchase the subscription.
5. Go to the edit subscription screen and from the actions drop down, process a renewal order. (see screenshot below).
6. Go to your Stripe dashboard and the transaction list. 
7. Note there are 2 or more charges for the same renewal order ID.

Screenshot 2023-02-27 at 1 27 38 pm

Admin Process renewal action.

### Actual behavior

For every payment method class that extends the `WC_Payment_Gateway_WCPay` class, there's a `$this->maybe_init_subscriptions()` call and that leads to multiple instances WC Payments hooking onto `woocommerce_scheduled_subscription_payment_`. 

For each one of those functions hooked onto that hook, there's a renewal charge being processed.

### Screenshots

Screenshot 2023-02-27 at 1 25 25 pm

Duplicate charges. 

### Expected behavior

A customer should only be charged once for each renewal order.

### PR DESCRIPTION
**Pull request by bborman22**
Fixes #5615 

#### Changes proposed in this Pull Request

This PR addresses the issue that was discovered in #5615. In short, the issue was that subscription renewals were running multiple charges for each renewal. It was discovered that it was running transactions for as many UPE payment methods that were available + 1. We discovered this was due to the fact that the `woocommerce_scheduled_subscription_payment_woocommerce_payments` action was being added for every payment method class. This was happening since each class was ultimately inheriting from `WC_Payment_Gateway_WCPay` which calls `maybe_init_subscriptions` in it's constructor and each class was calling it's parent contructor within it's own constructor.

To address this issue we took two steps, one to explicity avoid this renewal duplicate charge issue and one to avoid any other oddities that might have been waiting in hiding due to the multiple calls to `maybe_init_subscriptions`. 

- Addressing the renewal duplicate charge issue explicitly involved checking for the existence of the `woocommerce_scheduled_subscription_payment_woocommerce_payments` before adding it. In this way we ensure that the action is only ever added once (and thus only ever called once).
- Avoiding any further oddites involved removing the `maybe_init_subscriptions` from the `WC_Payment_Gateway_WCPay` constructor and instead calling it explicitly once all the of the enabled / relevant gateways have been instantiated.

#### Testing instructions
The following test should all be performed under each of these four scenarios:
- No UPE at all.
- Legacy UPE.
- Split UPE without WooPay.
- Split UPE with WooPay.

1. Configure store for the current scenario you are testing above.
2. If not already present, install the WC Subscriptions plugin and add a subscription product.
3. Go to checkout a subscription product using a CC.
4. Confirm subscription checkout was successful and a subscription was created.
5. Now as merchant admin, go into the subscription that was created and under the subscription actions, choose "Process Renewal" and hit Update.
6. Confirm that the renewal order was created and paid under the subscription order page under "Related Orders".
7. Log into the Stripe dashboard and view the test merchant account's payments and confirm that there is only one charge for the renewal.

After confirmation of the above, run an additional test with a subscription and SEPA.
1. Enable SEPA on your merchant account from the Stripe dashboard.
2. Restart your local WCPay server memcache instance to clear cached account data.
3. Refresh account data on merchant store with the dev tools.
4. Under Payment settings, make sure SEPA is enabled.
5. Checkout with a subscription product and repeat steps 5, 6, and 7 from above.

**Note:** While the scenarios above will cover the primary expected areas of impact, it wouldn't hurt to confirm some processes with WooPay enabled and disabled in non-UPE and legacy UPE as well.

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

### CODE REVIEW

File: includes/compat/subscriptions/trait-wc-payment-gateway-wcpay-subscriptions.php
Comment (c-shultz) [inline includes/compat/subscriptions/trait-wc-payment-gateway-wcpay-subscriptions.php:137]: Having a safety check here to prevent multiple renewals seems really sensible one way or another.

That being said, how do you all feel about the current state of the UPE implementation that caused the extra call in the first place. Do we also need to fix the originating issue to bring this to a satisfactory fix?

Comment (bborman22) [reply includes/compat/subscriptions/trait-wc-payment-gateway-wcpay-subscriptions.php:137]: I'll defer to @FangedParakeet and others that are more familiar with the recent UPE work, but I think to root out the originating issue might be a large / regression prone task. My understanding is that we need to modify how / when we're calling the parent gateway (`WC_Payment_Gateway_WCPay`) constructor as it is not getting the UPE payment method IDs to properly name hooks, etc. However, this change would only really apply to Split UPE as I don't believe non-UPE and legacy UPE rename the methods at all. For a rough idea of what that might look like, it was [initially looked into here](https://github.com/Automattic/woocommerce-payments/pull/5530/commits/889b224032ff615d66715f21421c99d3a3b6d6ad), but in a quick test today that seemed to cause some issues in looking up the subscription (though I didn't put much time into it beyond that).

Comment (bborman22) [reply includes/compat/subscriptions/trait-wc-payment-gateway-wcpay-subscriptions.php:137]: @FangedParakeet's update makes this a much more thorough solution by avoiding the multiple calls to `maybe_init_subscriptions`. That change has now been rolled into this PR as well for a two pronged solution.

**Review by FangedParakeet — COMMENTED** [review]
✅ Code LGTM. Tested with both CC and SEPA renewal orders and was unable to generate any duplicate charges. These latest changes should ensure that hooks are not duplicated and we have an additional guard to ensure that the scheduled subscription payment hook, that creates charges, cannot be duplicated for a payment gateway.

I would approve this PR--and do so in spirit--but since I have now contributed considerably to its contents, I think it would be more appropriate if someone else could deliver the official stamp of approval. 🙏

**Comment by reykjalin** [comment]
Before starting my review I think we need to test both "No UPE at all, without WooPay" and "No UPE at all, with WooPay" to fully cover everything? 🤔 

Maybe we need WooPay enabled/disabled variations for all cases?

File: includes/compat/subscriptions/trait-wc-payment-gateway-wcpay-subscriptions.php
Comment (hsingyuc) [reply includes/compat/subscriptions/trait-wc-payment-gateway-wcpay-subscriptions.php:137]: If we check `has_action` before running this hook, how will this work if someone else also used this hook?

For example, if another plugin uses this action first:

```
add_action( 'woocommerce_scheduled_subscription_payment_' . $id, function() {} );
```

Then our `scheduled_subscription_payment` would not execute.

**Comment by bborman22** [comment]
@reykjalin I would definitely say the more scenarios we cover the better. The reason for the original list is that WooPay's enablement should only have an impact on Split UPE's gateways. However, getting extra confirmation is always good. I'll add a note about additional testing with different arrangements of WooPay enabled.

File: includes/class-wc-payments.php
Comment (hsingyuc) [inline includes/class-wc-payments.php:478]: I think this fixes the issue, but is there a way we can check if this has loaded inside of `maybe_init_subscriptions()`?

**Comment by reykjalin** [comment]
With Split UPE and WooPay enabled it seems I can't checkout with a subscriptions product:

image

image

This isn't a problem on `develop` so I assume it's something introduced in this PR 🤔

**Comment by reykjalin** [comment]
The issue has something to do with the `maybe_init_subscriptions` call being moved out of the `WC_Payment_Gateway_WCPay` constructor, so probably just not being called on the right gateway where we moved it.

If I add it back to the gateway constructor I can pay, but that re-introduces the issue we're trying to fix so I think we just need to call the function on a different gateway where we moved the call? Will dig further.

File: includes/compat/subscriptions/trait-wc-payment-gateway-wcpay-subscriptions.php
Comment (FangedParakeet) [reply includes/compat/subscriptions/trait-wc-payment-gateway-wcpay-subscriptions.php:137]: I'm not saying this is impossible, but someone else would have to register the action `woocommerce_scheduled_subscription_payment_woocommerce_payments` in order to overwrite our own hook here. From our perspective, our goal is to ensure that this hook only fires once for any WCPay Subscription payment process. Ensuring that `maybe_init_subscriptions` only fires once, by calling it in a single place in `WC_Payments::init()` should also accomplish this, but this check is just to be twice as certain. 

Let us know if you think this is redundant or if there is a better way this could be handled!

**Comment by james-allan** [comment]
I note that you've all spent a lot more time testing this and considering the changes. I've mostly focused on these changes from a subscriptions perspective. Below is the set of setup/environments that I tested: 

| Setup        | `trunk`       | This branch|
|:------------:|:-------------:|:----------:|
| **Split UPE**    | 11 renewal charges 🔴  | 👍         |
| **Legacy UPE**   | 2 renewal charges 🔴      |   👍       |
| **No UPE**       |    👍    |    👍       |

<sup>👍 - 1 renewal charge per order</sup>

I've also tested other subscription related flows on this branch. 

| Flow        | This branch       |
|:------------:|:-------------:|
| **Subscription purchase**| 👍 |
| **renew early**| 👍 |
| **renew early via modal**| 👍 |
| **suspend/activate**| 👍 |
| **WC Payments subscription purchase** |    👍   |
| **WC Payments subscription cancelled** |    👍   |

Based on the changes I wouldn't expect any of these flows to be broken but wanted to double check given the number of hooks attached in the `maybe_init_subscriptions()`.


--- 

### Comments 

The changes look good and align with my understanding of the issues. 

I think there is a need to revisit this section of subscription integration code and tidy up it's intention. All instances of payment method classes include the `WC_Payment_Gateway_WCPay_Subscriptions_Trait` via the base `WC_Payment_Gateway_WCPay` class. This trait appears to be more akin to a wc payments plugin wide integration class rather than a set of functions that all payment method instances need to inherit natively.

If at some point in the future split UPE changes the `$this->id` or the order of operations, there are these hidden and deep level consequences that I don't think the developer would be aware of. If the intent is for 90% of this trait to only run once, than they should be moved out to a global integration/compatibility class rather than sit in this trait. 

## ❓ Question

**As an extra layer of defence, what about if we changed `scheduled_subscription_payment()` to also check if the order needs payment?**

From my understanding, part of the original issue is that this function was hooked in 8 or 9 times in some situations but also those 9 callbacks all processed a payment when after the first payment, there would have been no need to.

**Review by reykjalin — APPROVED** [review]
I'm approving this since the only case where the fix causes problems is when merchants have access to Split UPE, have Split UPE enabled, have access to WooPay, and have WooPay enabled. That segment of merchants is probably going to be 0 for the foreseeable future given that WooPay isn't generally available to merchants yet.

It's more important to enable renewal payments ASAP for the rest of our merchants. That said I'd like to see us start to work on a refactor that properly addresses the problem immediately after this has been merged.

File: includes/compat/subscriptions/trait-wc-payment-gateway-wcpay-subscriptions.php
Comment (james-allan) [reply includes/compat/subscriptions/trait-wc-payment-gateway-wcpay-subscriptions.php:137]: > If we check has_action before running this hook, how will this work if someone else also used this hook?

I think this is a good point and introducing that risk is something we should keep in mind.

While the chance is super low, it is possible. What makes it even more complicated is that if this were to happen, it would be pretty hard to discover that is the cause.

**Comment by reykjalin** [comment]
I should also mention the configurations I tested like James did:

On `develop`:

| | No UPE | Legacy UPE enabled | Split UPE enabled |
|:---:|:---:|:---:|:---:|
| **WooPay enabled** | ✅ | 2 renewals | 11 renewals |
| **WooPay disabled** | ✅ | 2 renewals | 11 renewals |

On this branch:


| | No UPE | Legacy UPE enabled | Split UPE enabled |
|:---:|:---:|:---:|:---:|
| **WooPay enabled** | ✅ | ✅ | [Failed](https://github.com/Automattic/woocommerce-payments/pull/5619#issuecomment-1447372453) because there is no gateway registered |
| **WooPay disabled** | ✅ | ✅ | ✅ |

**Comment by hsingyuc** [comment]
I think adding a property `private static $has_subscriptions_initialized` to the traits might be an option that does not introduce duplicate calls to these hooks and doesn't possibly break other hooks.

```
trait WC_Payment_Gateway_WCPay_Subscriptions_Trait {
	private static $has_subscriptions_initialized = false;

	public function maybe_init_subscriptions() {
		if ( ! self::$has_subscriptions_initialized ) {
			add_action( 'woocommerce_scheduled_subscription_payment_' . $this->id, [ $this, 'scheduled_subscription_payment' ], 10, 2 );
		}

		self::$has_subscriptions_initialized = true;
	}
}
```

We can use the `has_action` check but this might create more bugs in the future.

**Comment by james-allan** [comment]
> With Split UPE and WooPay enabled it seems I can't checkout with a subscriptions product:
> 
> The issue has something to do with the maybe_init_subscriptions call being moved out of the WC_Payment_Gateway_WCPay constructor, so probably just not being called on the right gateway where we moved it.

@reykjalin Is it possible that this is because inside the `maybe_init_subscriptions()` function is this line:

```php
 $this->supports = array_merge( $this->supports, $payment_gateway_features );
```

Prior to this PR, every payment method object called the `maybe_init_subscriptions()` function and it would add subscription-related feature support to those payment methods. 

By moving `maybe_init_subscriptions()` and calling it only on the card gateway instance, only the card method now has subscription related features.

🤔 

I don't have much experience to understand how WooPay works but the error message included in your screenshot makes me think it is a subscription-support flag related issue.

**Comment by james-allan** [comment]
> @reykjalin Is it possible that this is because inside the maybe_init_subscriptions() function is this line:
>
> ```
> $this->supports = array_merge( $this->supports, $payment_gateway_features );
> ```

Taking a step back, when I look at `maybe_init_subscriptions()` it can be categorised in 2 parts:

1. the `$payment_gateway_features` section which conditionally adds subscription related features to the gateway object depending on whether the subscriptions plugin is active or if subscriptions is enabled via the free feature in WC Payments.
2. Attaching action/filter callbacks. 

I can't see how attaching the callbacks would make any difference in terms of the payment method availability error you saw so I can only imagine the issue you're seeing has to be related to the `$supports` handling. I could be wrong though.

In any case, I think there might be another approach available and it make sure of the feedback provided by @hsingyuc. 

`maybe_init_subscriptions()` is doing more than just attaching callbacks which is at the root of the duplicate renewal issue so to patch the root problem, I suggest we:

1. Return the `maybe_init_subscriptions()` to where it was - inside the `WC_Payment_Gateway_WCPay` constructor. This will restore the `$supports` handling.
2. Inside `maybe_init_subscriptions()` introduce a check that @hsingyuc suggested so that we only attach the callbacks once, but leave the rest of the function to run.
3. Remove the condition which checks `has_action()`. While the risk is small, it is introducing a risk that would lead to renewal order attempts not being made at all and given we're at the point of patching something I'd suggest we err on the side of caution. 
4. As an extra layer of security, change `scheduled_subscription_payment` so it doesn't process a payment for a renewal order that is already paid. 

In the interests of keeping this patch moving, I'm going to submit a separate PR for review.

If the consensus is that folks are happier with this approach, then that's fine.

**Comment by RadoslavGeorgiev** [comment]
I'm yet to fully review/test the PR, but being the patch it is, let's change the base branch to `trunk`. There's a lot of stuff in `develop`, which should not be released together with this yet.

**Comment by james-allan** [comment]
> I'm yet to fully review/test the PR, but being the patch it is, let's change the base branch to trunk. There's a lot of stuff in develop, which should not be released together with this yet.

Based on this comment in Slack, p1677565463476069-slack-C04RDE70TUJ, it appears that all the other PRs we intend on releasing in this patch (see below) have been merged into develop. 

- https://github.com/Automattic/woocommerce-payments/pull/5610
- https://github.com/Automattic/woocommerce-payments/pull/5607
- https://github.com/Automattic/woocommerce-payments/pull/5614
- https://github.com/Automattic/woocommerce-payments/pull/5594

I'm not sure why that approach was taken, and what implications it has for this PR or the release process.

**Comment by reykjalin** [comment]
> If the consensus is that folks are happier with this approach, then that's fine.

As you noted in Slack, we basically arrived at the same conclusion! Just noting that here so the discussion is completed in both places 🙂

**Comment by shendy-a8c** [comment]
It's OK. I'll cherry-pick once this PR (or https://github.com/Automattic/woocommerce-payments/pull/5625) is merged to develop.

**Comment by shendy-a8c** [comment]
Hi, since this PR has not yet merged, I wonder if it's OK to change the PR's base branch to `trunk`. Discussion: p1677579853660769/1677578222.590459-slack-CGGCLBN58.

**Comment by shendy-a8c** [comment]
To whoever that will merge this PR, please read my comment https://github.com/Automattic/woocommerce-payments/pull/5625#issuecomment-1448077919.

**Comment by bborman22** [comment]
Closing this PR in favor of #5635

**Comment by c-shultz** [comment]
>Closing this PR in favor of https://github.com/Automattic/woocommerce-payments/pull/5635

Quick correction here for anyone looking back at history on this issue -- this should be #5625

**Pull request closed without merging**
